sql-tuning-best-practice.md

```diff
@@ -0,0 +1,952 @@
+|       └─HashAgg_13                         | 1.00        | 1        | mpp[tiflash] |                | tiflash_task:{time:1.41s, ...        |
+|         └─Projection_74                    | 3813443.11  | 3864397  | mpp[tiflash] |                | tiflash_task:{time:1.4s, ...         |
+|           └─Projection_51                  | 3813443.11  | 3864397  | mpp[tiflash] |                | tiflash_task:{time:1.39s, ...        |
+|             └─HashJoin_50                  | 3813443.11  | 3864397  | mpp[tiflash] |                | tiflash_task:{time:1.39s, ...        |
+|               ├─ExchangeReceiver_31(Build) | 3800312.67  | 3864397  | mpp[tiflash] |                | tiflash_task:{time:1.05s, ...        |
+|               │ └─ExchangeSender_30        | 3800312.67  | 3864397  | mpp[tiflash] |                | tiflash_task:{time:1.2s, ...         |
+|               │   └─TableFullScan_28       | 3800312.67  | 3864397  | mpp[tiflash] | table:lineitem | tiflash_task:{time:1.15s, ...        |
+|               └─ExchangeReceiver_34(Probe) | 10000000.00 | 10000000 | mpp[tiflash] |                | tiflash_task:{time:1.24s, ...        |
+|                 └─ExchangeSender_33        | 10000000.00 | 10000000 | mpp[tiflash] |                | tiflash_task:{time:1.4s, ...         |
+|                   └─TableFullScan_32       | 10000000.00 | 10000000 | mpp[tiflash] | table:part     | tiflash_task:{time:59.2ms, ...       |
++--------------------------------------------+-------------+----------+--------------+----------------+--------------------------------------+
+```
+
+### SaaS Arbitrary Filtering Workloads
+
+#### Overview
+
+In SaaS applications, it's common to structure tables with composite primary keys that include tenant identification. Let's look at a typical example where TiFlash can significantly improve query performance.
+
+#### Case Study: Multi-tenant Data Access
+
+Consider a table design with a composite primary key: (tenantId, objectTypeId, objectId). A common query pattern is to:
+
+- Fetch the first N records for a specific tenant and object type, and applying random filters across hundreds or thousands of other columns, Making it impractical to create indexes for every possible filter combination. There is a potential sort operator after the filter as well.
+- Get the total count of records matching these criterias
+
+#### Performance Comparison
+
+- TiKV Plan: When executing this query on TiKV storage engine, it takes 2 minutes 38.6 seconds as it requires a table range scan operation, and cannot utilize indexes effectively for dynamic filtering conditions across hundreds of columns.
+- TiFlash Plan: The same query on TiFlash MPP engine takes only 3.44 seconds - almost 46 times faster. Since data in TiFlash is sorted by primary key, queries filtered by the primary key's prefix will also use a TableRangeScan instead of a full table scan.
```

Comment:
It's wrong. Both plans don't contain a full table scan.

From the limited information, I think there are 2 main reasons for the performance difference:
1. Data is spread over 5000+ regions for tikv (which implies 5000+ cop tasks), but tiflash only uses 2 mpp tasks. Besides the RPC overhead, it also means the mpp plan provides a better aggregation effect in the 1st stage aggregation and less aggregation work to do.
2. TiFlash scans and aggregates much faster than TiKV.

Reply:
Would it not be 5000+ regions in TiFlash as well, but for both TiKV and TiFlash it can use the `TenantID` to only use matching regions? (like range partitioning, but directly on the clustered index by regions). But of course RPC overhead may come in to play as well.

Reply:
Yes. I'm not familiar with the implementation of TiFlash, but I believe the data is not stored by regions in TiFlash, and it also doesn't need to split the mpp task according to regions in TiFlash. There only needs to be 1 mpp task sent to 1 TiFlash node for each table, and it's executed as a whole.